uclasm/matching/matching_problem.py

```diff
@@ -274,3 +277,46 @@ def __str__(self):
                              .format(n_missing, missing_ground_truth))
 
         return "\n".join(info_strs)
+
+    def reduce_world(self):
+        """Reduce the size of the world graph.
+
+        Check whether there are any world nodes that are not candidates to
+        any tmplt nodes. If so, remove them from the world graph and update
+        the matching problem.
+
+        Returns
+        -------
+        bool
+            True if the size of the world is reduced. False otherwise.
+        """
+        # Note: need to update the global_costs before reduce_world to reflect
+        # changes in the candidates
+
+        is_cand = self.candidates().any(axis=0)
+
+        # If some world node does not serve as candidates to any tmplt node
+        if ~is_cand.all():
+            self.world = self.world.node_subgraph(is_cand)
+            self.shape = (self.tmplt.n_nodes, self.world.n_nodes)
+
+            # Update parameters based on new world
+            self.set_costs(local_costs=self.local_costs[:, is_cand])
+            self.set_costs(fixed_costs=self.fixed_costs[:, is_cand])
+            from_local_bounds(self)
```

Comment:
After changing from_local_bounds back to monotone array comparison, add a line before local bounds to set_costs so that the array is appropriately reshaped.